### PR TITLE
feat(babel-parser): throw syntax error for `}` and `>` in JSX text

### LINKED ISSUE #11042: Proposal: throw syntax error for `}` and `>` in JSX text.
## Feature Request

- [x] I would like to work on this feature!
  - If needs be, I can look into it

**Is your feature request related to a problem? Please describe.**

The JSX spec [1][2] lists `}` and `>` as invalid JSX text characters.

> JSXTextCharacter :
> - SourceCharacter but not one of {, <, > or }

However, babel does not throw a syntax error when it encounters a naked `}` or `>` in a JSXText string.

According to the spec, the only way you should be able to include either of these characters (similarly for `{` and `<`), is via a string expression: `{'>'}`.

**Describe the solution you'd like**
I propose that babel should follow the JSX spec, and throw a syntax error if it encounters a "naked" `}` or `>` within jsx text.

From my experience, it's uncommon that these characters are intentionally put in JSX text, and are instead the result of typos (i.e. typing `{}`, but having an editor auto-complete an extra `}` => `{}}`), or copy pasting errors.

Flow has already updated their parser to flag this as an error [3][4]. TypeScript is happy to make this change as well [5] (if Babel makes the change, as babel is generally considered the "reference implementation" for JSX).
I have also raised an issue in acorn for the same functionality [6].

**Describe alternatives you've considered**
- Custom lint rules
  - We had these implemented at facebook before flow made it a parsing error. (The lint rules actually spurred the change).
  - Outside of a tightly controlled monorepo, I can see that it would be hard to gain adoption for the rule, as not everyone uses linting tools, not everyone uses the react plugins, etc etc.
- Prettier auto-formatting to convert `}` to `{'}'}`.
  - This is an AST change, and prettier doesn't like to change the AST when formatting.

**Teachability, Documentation, Adoption, Migration Strategy**
If you can, explain how users will be able to use this and possibly write out a version the docs.
Maybe a screenshot or design?

I think the best way to document this would be via clear and explicit error messaging.
For example, the following code:
```js
function F() {
  return <div>></div>;
}

function G() {
  return <div>{1}}</div>;
}
```
produces the following flow errors:
```
4:   return <div>></div>;
                 ^ Unexpected token `>`. Did you mean `{'>'}`?
8:   return <div>{1}}</div>;
                    ^ Unexpected token `}`. Did you mean `{'}'}`?
```
[flow.org repl](https://flow.org/try/#0MYewdgzgLgBASgUwIbFgXhgJwQRwK4CW2AFAOTYpSkCUA3AFD0BmeYqB4MAYsdTAN70YWBFDyYwMADwATAgDcAfIqkB6OUoYBfRizZQOkgOK8BQkWInSNi-gEYtWtTe1A)

---

Related:
[1] https://facebook.github.io/jsx/
[2] https://github.com/facebook/jsx/pull/18
[3] https://github.com/facebook/flow/blob/master/Changelog.md#01160
[4] facebook/flow@e1d0038
[5] https://github.com/microsoft/TypeScript/issues/36341
[6] https://github.com/acornjs/acorn-jsx/issues/106

### PR DESCRIPTION
**Pull request by bradzacher**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #11042
| Patch: Bug Fix?          |
| Major: Breaking Change?  | Yes
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

See the full context in #11042.
The JSX spec lists `}` and `>` as invalid JSX text characters.
Flow recently updated their parser so that it errors in this case.
This PR just does the same for Babel.

Looking for feedback, as I've never contributed to babel before.
Do we need explicit documentation for this change, or are the error messages "good enough"?

### CODE REVIEW

File: packages/babel-parser/src/plugins/jsx/index.js [hunk 1]
```diff
@@ -507,13 +507,35 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return this.jsxParseElementAt(startPos, startLoc);
     }
 
+    // Validates and parses the current state as JSXText
+
+    parseJSXText(): N.Expression {
+      const startPos = this.state.start;
+      for (let idx = 0; idx < this.state.value.length; idx += 1) {
+        const character = this.state.value[idx];
+        if (character === "}") {
```
Comment (kaicataldo) [inline packages/babel-parser/src/plugins/jsx/index.js]: Do we also need to need to throw an error for `{` and `<`?

File: packages/babel-parser/src/plugins/jsx/index.js [hunk 2]
```diff
@@ -507,13 +507,35 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+    }
+
     // ==================================
     // Overrides
     // ==================================
 
     parseExprAtom(refExpressionErrors: ?ExpressionErrors): N.Expression {
       if (this.match(tt.jsxText)) {
-        return this.parseLiteral(this.state.value, "JSXText");
+        return this.parseJSXText();
```
Comment (jridgewell) [inline packages/babel-parser/src/plugins/jsx/index.js]: It seems I pointed you to the wrong location, sorry about that!

At this point, we've already tokenized the source into a `tt.jsxText` token (that's the `this.match()` check above). But we can do this earlier, when we're still creating the `jsxText` token. It's located at https://github.com/babel/babel/blob/1815c349877acca1d3275ea72512ee7824db49e3/packages/babel-parser/src/plugins/jsx/index.js#L81-L120.

File: packages/babel-parser/src/plugins/jsx/index.js (hunk 1 above)
Comment (kaicataldo) [reply packages/babel-parser/src/plugins/jsx/index.js]: Ah, ignore me - this already throws.

File: packages/babel-parser/src/plugins/jsx/index.js (hunk 2 above)
Comment (bradzacher) [reply packages/babel-parser/src/plugins/jsx/index.js]: ezpz I'll fix that up tomorrow morning.
thanks for the pointers!

Comment (bradzacher) [reply packages/babel-parser/src/plugins/jsx/index.js]: ezpz I'll fix that up tomorrow morning.
thanks for the pointers!

Comment (bradzacher) [reply packages/babel-parser/src/plugins/jsx/index.js]: ezpz I'll fix that up tomorrow morning.
thanks for the pointers!

File: packages/babel-parser/src/plugins/jsx/index.js [hunk 3]
```diff
@@ -107,6 +107,15 @@ export default (superClass: Class<Parser>): Class<Parser> =>
             chunkStart = this.state.pos;
             break;
 
+          case charCodes.greaterThan:
+          case charCodes.rightCurlyBrace:
+            throw this.raise(
+              this.state.pos,
+              `Unexpected token \`${
+                this.input[this.state.pos]
+              }\`. Did you mean \`{'${this.input[this.state.pos]}'}\`?`,
```
Comment (jridgewell) [inline packages/babel-parser/src/plugins/jsx/index.js]: Nit: we should encourage the HTML entities `&rbrace;` and `&gt;`, which shouldn't require the separate `JSXExpressionContainer`.

Comment (bradzacher) [reply packages/babel-parser/src/plugins/jsx/index.js]: Would you be okay with the message suggesting both? i.e.
``Unexpected token `>`. Did you mean `{'>'}`, or `&gt;`?``

Personally I don't like seeing html entity codes in JSX.

Comment (jridgewell) [reply packages/babel-parser/src/plugins/jsx/index.js]: Sure.

**Comment by nicolo-ribaudo** [comment]
Thanks! It will take a bit to merge this PR since we need to wait for Babel 8.

File: packages/babel-parser/src/plugins/jsx/index.js [hunk 4]
```diff
@@ -107,6 +107,20 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
 
+          case charCodes.greaterThan:
+          case charCodes.rightCurlyBrace: {
+            const htmlEntity =
+              ch === charCodes.rightCurlyBrace ? "&rbrace;" : "&gt;";
+            throw this.raise(
+              this.state.pos,
+              `Unexpected token \`${
+                this.input[this.state.pos]
+              }\`. Did you mean \`${htmlEntity}\`, or \`{'${
```
Comment (kaicataldo) [inline packages/babel-parser/src/plugins/jsx/index.js]:
```suggestion
              }\`. Did you mean \`${htmlEntity}\` or \`{'${
```

Small thing I just noticed: Can we get rid of this extra comma in the error message?

**Comment by nicolo-ribaudo** [comment]
I'm merging this to the `next-8-dev` branch. It will take some time before it's released!

**Pull request merged**